### PR TITLE
Bugfix/de462464 unsupported entities

### PR DESCRIPTION
**Pull request by uppoju**
## Description
Unsupported entity type is added to support all unsupported entity types in dev plugin
## Links
Link Type   | Link
------      | ------
Rally Issue | de462464 
Func Spec   | 
## Checklist
- [x] Pull Request Created
- [ ] Code Review Completed
- [ ] Veracode scan results addressed
- [ ] SonarQube scan results addressed
- [ ] TPSR has been submitted (if applicable) 
- [ ] Unit tests have been added
- [ ] Integration/Functional test cases have been written and automated
- [ ] Func Spec has been updated as needed
- [ ] Rally Issue(s) are in an engineering completed state

### CODE REVIEW

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+public class UnsupportedEntityFilter implements EntityFilter<UnsupportedGatewayEntity> {
+    @Override
+    public @NotNull Collection<Class<? extends EntityFilter>> getDependencyEntityFilters() {
+        return Collections.EMPTY_LIST;
+    }
+
+    @Override
+    public List<UnsupportedGatewayEntity> filter(String folderPath, FilterConfiguration filterConfiguration, Bundle bundle, Bundle filteredBundle) {
+        return bundle.getEntities(UnsupportedGatewayEntity.class).values().stream()
+                .filter(p ->! p.getType().equals("RBAC_ROLE")).collect(Collectors.toList());
```
Comment (uppoju) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java:23]: there are so many RBAC_ROLE entities getting created in unsupported-entities.yml, do we have to skip ?

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/UnsupportedEntityLoader.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    private final EntityUtils.GatewayEntityInfo gatewayEntityInfo;
+
+    @Inject
+    public UnsupportedEntityLoader(final JsonTools jsonTools, final IdGenerator idGenerator) {
+        super(jsonTools, idGenerator);
+        this.gatewayEntityInfo = EntityUtils.createEntityInfo(UnsupportedGatewayEntity.class);
+    }
+
+    @Override
+    public void load(final Bundle bundle, final File rootDir) {
```
Comment (RaghuReddy) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/UnsupportedEntityLoader.java:38]: @uppoju , do we need this load method, env entities loads with below load method, also here folder is hard coded to config directory

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java [hunk 3]
```diff
@@ -324,19 +324,32 @@ private void loadFolderDependencies(AnnotatedBundle annotatedBundle, GatewayEnti
...
-                            if (gatewayEntity.getName() != null) {
-                                return dependency.getName().equals(gatewayEntity.getName());
-                            } else {
-                                return dependency.getName().equals(PathUtils.extractName(e.getKey()));
-                            }
-                        }).findFirst();
-        Map entityMap = annotatedBundle.getEntities(entityClass);
-        optionalGatewayEntity.ifPresent(e -> entityMap.put(e.getKey(), e.getValue()));
+        if(entityClass != null){
+            Map<String, ? extends GatewayEntity> allEntitiesOfType = rawBundle.getEntities(entityClass);
```
Comment (RaghuReddy) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:328]: @uppoju , can you please change the rawBundle name to fullBundle in this PR, it is actually confusing with actual rawBundle which we get at the time of export.

Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java:328]: We dont get raw bundle out of export, its a resultant bundle

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/UnsupportedEntityLoader.java (hunk 2 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/UnsupportedEntityLoader.java:40]: Unsupported entities have xml along with yml file, super.load methods loads only yml file
Even for other environment entities jsonTools.getDocumentFile(rootDir, getFileName()) loads from config folder only

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java (hunk 1 above)
Comment (graju256) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java:23]: I think YES for now.

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/UnsupportedEntityWriter.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public void write(Bundle bundle, File rootFolder, Bundle rawBundle) {
+        File configFolder = new File(rootFolder, "config");
+        documentFileUtils.createFolder(configFolder.toPath());
+
+        Map<String, UnsupportedGatewayEntity> unsupportedGatewayEntityMap = bundle.getUnsupportedEntities();
+        DocumentBuilder builder = documentTools.getDocumentBuilder();
+        Document document = builder.newDocument();
+        Element items = document.createElement("l7:Items");
+        unsupportedGatewayEntityMap.values().parallelStream().forEach(unsupportedGatewayEntity -> {
```
Comment (graju256) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/UnsupportedEntityWriter.java:44]: As long as you are dealing with single file, you don't get any benefit of using parallel streams. Moreover, it creates some unknown issues. Do it a normal way.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/environment/BundleEnvironmentValidator.java [hunk 5]
```diff
@@ -103,7 +104,11 @@ private void findInBundle(Bundle bundle, String type, String name) {
                 entity = bundle.getJmsDestinations().get(name);
                 break;
             default:
-                throw new MissingEnvironmentException("Unexpected entity type " + type);
+                UnsupportedGatewayEntity unsupportedGatewayEntity = bundle.getUnsupportedEntities().get(name);
+                if(unsupportedGatewayEntity != null && type.equals(unsupportedGatewayEntity.getType())){
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/environment/BundleEnvironmentValidator.java:108]: Can you write something to log at INFO or WARNING level?

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java (hunk 1 above)
Comment (graju256) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java:23]: Please skip them. Are there any more that we need to skip?

**Review by RaghuReddy — APPROVED** [review]
LGTM

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/environment/BundleEnvironmentValidator.java (hunk 5 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/environment/BundleEnvironmentValidator.java:108]: Done

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java (hunk 1 above)
Comment (uppoju) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/filter/entityfilters/UnsupportedEntityFilter.java:23]: Added  a few more entity types

**Review by graju256 — APPROVED** [review]

**Pull request merged**